common_arm/usb_cdc.h

```diff
@@ -30,6 +30,8 @@ bool usb_poll(void);
 bool usb_poll_validate_length(void);
 uint32_t usb_read(uint8_t *data, size_t len);
 int usb_write(const uint8_t *data, const size_t len);
+void usb_write_byte_async(uint8_t data);
```

Comment:
First, I admit I may be fully misunderstanding something.

<details><summary>These functions appear to have implicit hidden relationship</summary><p/>

It seems that these two new functions have an implicit relationship, which is not obvious from their names.
Specifically, it would appear that the following sequence is necessary:

1. At function entry, clearing any prior transfers / data in USB FIFO buffer.  No function for this currently exists.
2. Calling `usb_write_byte_async()` ... which really just adds a byte to the hardware FIFO for the next transfer?
3. Tracking how many bytes are in the hardware FIFO buffer (e.g., how many times called above function....)
4. Before hardware FIFO fills, call `usb_write_request()` ... which may spin / cause lost samples ....
5. At function exit, if hardware FIFO is not empty, send final (partial) packet of data to empty FIFO buffer.
    (no function exists for this at the moment?)

</details> 

<details><summary>Naive code added later based on current names</summary><p/>

To show how the names may suggest functionality, such as presuming that async writes would eventually send properly, but result in really unexpected and difficult-to-parse problems...

```
// send buffer contents twice
for (int i = 0; i < 512; i++) {
    usb_write_byte_async(buffer[i]); // definitely does not result in sending 512 bytes
}
usb_write_request(); // b/c hardware FIFO is < 512 bytes (328 bytes ... likely split across all endpoints?)
for (int i = 0; i < 512; i++) {
    // maybe user doesn't even recognize need to call usb_write_request()...
    usb_write_byte_async(buffer[i]); 
}
usb_write(buffer, 512); // who knows what odd results will happen here?
usb_write_byte_async(0x77); // this byte might be lost, or corrupt the next packet?
usb_write(buffer, 512);
usb_write_request(); // what would this do?
```
</details> 

### Questions for discussion....

1. How can this be structured to make clear the hidden bindings between these functions?
    1. For example, that what is currently named `usb_write_byte_async()` requires a call to what is currently named `usb_write_request()`, before other calls that write to USB?
1. Could the answer be as simple as renaming the functions and providing inline function documentation?
1. What happens if too many bytes are attempted to be written asynchronously?
    1. Does the CPU hang?
    1. Is the last byte discarded?
    1. Is the first byte in the FIFO discarded?
    1. Other?

Reply:
> which really just adds a byte to the hardware FIFO for the next transfer?
Yes.

I admit that the name of these functions are somehow confusing. These inline functions are just hard-coded before 045a10cb03d060d2448d0e2071e2514adcfb2226, but as @iceman1001 suggested I extract them to a different .c file. I didn't select the name carefully. Would you like to help me to pick proper names for them?

Reply:
> What happens if too many bytes are attempted to be written asynchronously? 

I haven't tested that.

Reply:
In the folder  `common_arm\`  you find the usb_cdc.c  / .h  files.   Where these functions belongs.

Reply:
> I admit that the name of these functions are somehow confusing. 
> I didn't select the name carefully. Would you like to help me to pick proper names for them?

I think you created much clearer names with commit https://github.com/RfidResearchGroup/proxmark3/commit/079d2c7e37e116e82010715185e56c13547feb72.  👍 

(https://github.com/RfidResearchGroup/proxmark3/blob/079d2c7e37e116e82010715185e56c13547feb72/common_arm/usb_cdc.h#L34-L37)